File: algorithms/0350.intersection_of_two_arrays_ii/intersection_of_two_arrays_ii.rs
use std::{collections::HashMap};//, hash::Hash, iter::FromIterator};
impl Solution {
    pub fn intersect(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut dict = HashMap::new();
        for &n in nums2.iter() {
            let counter = dict.entry(n).or_insert(0);
            *counter += 1;
        }
        let mut result = vec![];
        for &n in nums1.iter() {
            if dict.get(&n) != None {
                result.push(n);
                *dict.get_mut(&n).unwrap() -= 1;
                if dict[&n] <= 0 {
                    dict.remove(&n);
                }
            }
        }
        result
    }
}

File: algorithms/0350.intersection_of_two_arrays_ii/intersection_of_two_arrays_ii.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> ans;
        unordered_map<int, int> num_counter;
        for (const auto& n: nums1) {
            ++num_counter[n];
        }
        for (const auto& n: nums2) {
            auto &counter = num_counter[n];
            if (counter > 0) {
                ans.push_back(n);
                --counter;
            }
        }
        return ans;
    }
};